ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt

```diff
@@ -200,32 +244,92 @@ internal class PaywallViewModelImpl(
+                PurchasesAreCompletedBy.MY_APP -> {
+                    checkNotNull(customPurchaseHandler) {
+                        "myAppPurchaseLogic must not be null when purchases.purchasesAreCompletedBy " +
+                            "is PurchasesAreCompletedBy.MY_APP"
+                    }
+                    when (val result = customPurchaseHandler.invoke(activity, packageToPurchase)) {
+                        is MyAppPurchaseResult.Success -> {
+                            val storeTransaction = result.purchase?.toStoreTransaction() ?: return
+                            val customerInfo = purchases.awaitCustomerInfo()
+                            listener?.onPurchaseCompleted(customerInfo, storeTransaction)
+                        }
+                        is MyAppPurchaseResult.Cancellation -> {
+                            trackPaywallCancel()
+                            listener?.onPurchaseCancelled()
+                        }
+                        is MyAppPurchaseResult.Error -> {
+                            val purchasesException = PurchasesException(result.error)
+                            throw purchasesException
+                        }
+                    }
+                }
+                PurchasesAreCompletedBy.REVENUECAT -> {
+                    if (customPurchaseHandler != null) {
+                        Logger.e(
+                            "myAppPurchaseLogic expected to be null " +
+                                "when purchases.purchasesAreCompletedBy is .REVENUECAT. \n" +
+                                "myAppPurchaseLogic.performPurchase will not be executed.",
+                        )
+                    }
+                    val purchaseResult = purchases.awaitPurchase(
+                        PurchaseParams.Builder(activity, packageToPurchase),
+                    )
+                    listener?.onPurchaseCompleted(purchaseResult.customerInfo, purchaseResult.storeTransaction)
+                }
+                else -> {
+                    Logger.e("Unsupported purchase completion type: ${purchases.purchasesAreCompletedBy}")
+                }
+            }
+
+            Logger.d("Dismissing paywall after purchase")
+            options.dismissRequest()
+        } catch (e: PurchasesException) {
+            if (e.code == PurchasesErrorCode.PurchaseCancelledError) {
+                trackPaywallCancel()
+                listener?.onPurchaseCancelled()
+            } else {
+                listener?.onPurchaseError(e.error)
+                _actionError.value = e.error
+            }
+        }
+
+        finishAction()
+    }
+
+    private fun validateState() {
+        if (purchases.purchasesAreCompletedBy == PurchasesAreCompletedBy.MY_APP && options.myAppPurchaseLogic == null) {
+            error(
+                "myAppPurchaseLogic is null, but is required when purchases.purchasesAreCompletedBy is .MY_APP.",
+            )
+        }
```

Comment:
The built-in `check` function has the same behavior, throwing `IllegalStateException`. Alternatively you can use `require` instead of `check`, which throws an `IllegalArgumentException`. Both can be considered correct in this case imo.
```suggestion
        check(
            purchases.purchasesAreCompletedBy != PurchasesAreCompletedBy.MY_APP || options.myAppPurchaseLogic != null
        ) { "myAppPurchaseLogic is null, but is required when purchases.purchasesAreCompletedBy is MY_APP." }
```  
  
   
I would also be fine to inline this in the constructor, as that's really the only place we should do this check. Having it as a separate function gives the impression that we might want to reuse this and call this in other places too. So you'd get: 

```kotlin
init {
    check(...) {...}
}
```

Reply:
I was aware of `check`/`require`, but I didn't find the boolean condition as easy to understand and immediately see as correct (or not), and `require` I thought threw the wrong exception type. So I'd leave this as-is, but will change it if you like!

I agree on moving it directly to the constructor, will do once I hear back on the above.

Reply:
> I didn't find the boolean condition as easy to understand and immediately see as correct (or not)

Could you clarify this? Is it because you'd be checking for the opposite condition?

Reply:
I just found the if condition here matches better how I think:

"if purchases are completed by my app, and my app's purchase logic does not exist, that's an error":

`purchases.purchasesAreCompletedBy == PurchasesAreCompletedBy.MY_APP && options.myAppPurchaseLogic == null`

vs 

"check that purchases are not completed by my app, or that my app's purchase logic exists, otherwise that's an error":

`check(purchases.purchasesAreCompletedBy != PurchasesAreCompletedBy.MY_APP || options.myAppPurchaseLogic != null)`. 

`check` is more concise and meant to do in one step what I've done in two, and normally should be used, but given the conditions that are being checked here I had to think harder to convince myself of the correctness of the code when using `check`. But maybe others find the `check` code fine to read and/or prefer it because it's more concise, in which case I'm happy to change it!

(This reminds me a bit of [this exchange](https://github.com/RevenueCat/purchases-ios/pull/3823#discussion_r1575364771) when choosing to use `if` when `guard` might have been technically preferred but not quite as readable.)

Reply:
Yea that's a good argument! The fewer brain cpu cycles we have to spend reading the code, the better. 

You could still do this ofc:
```kotlin
check(!(purchases.purchasesAreCompletedBy == PurchasesAreCompletedBy.MY_APP && options.myAppPurchaseLogic == null))
```
But I can agree to the `if` being clearer, because now we have a sneaky `!` in there that's easy to miss.

Reply:
What do you think about my suggestion to inline it in the constructor? So in case of the `if`:

```kotlin
init {
    if (...) {
        error(...)
    }
}
```